src/main/java/seedu/address/logic/commands/DeleteStudentCommand.java

```diff
@@ -44,14 +71,56 @@ public DeleteStudentCommand(Name name) {
     @Override
     public CommandResult execute(Model model) throws CommandException {
         requireNonNull(model);
+        ObservableList<Student> studentList = model.getFilteredStudentList();
+        ArrayList<Student> listToCheck = new ArrayList<>();
+
+        for (Student student : studentList) {
+            if (student.getName().equals(name)) {
+                listToCheck.add(student);
+            }
+        }
+
+        if (listToCheck.isEmpty()) {
+            throw new CommandException(MESSAGE_NONEXISTENT_STUDENT);
+        }
+
+        // if there are multiple students with the same name, ensure that student number is provided
+        if (listToCheck.size() > 1 && this.studentNumber == null) {
+            StringBuilder duplicates = new StringBuilder();
+            for (Student student : listToCheck) {
+                duplicates.append(student.getStudentNumber());
+                duplicates.append(" ");
+            }
+            String duplicateStudentNumbers = duplicates.toString();
+            throw new CommandException(String.format(MESSAGE_DUPLICATE_STUDENT, duplicateStudentNumbers, name));
+        }
+
+        if (listToCheck.size() > 1) {
+            for (Student student : listToCheck) {
+                if (student.getStudentNumber().equals(studentNumber)) {
+                    model.deleteStudent(student);
+                    return new CommandResult(String.format(MESSAGE_DELETE_STUDENT_SUCCESS) + student.getName()
+                            + " " + student.getStudentNumber());
+                }
+            }
+            throw new CommandException(MESSAGE_NONEXISTENT_STUDENT);
+        }
+
+        // happy path if exactly 1 student with the name is found
         Student studentToDelete = model.getStudentByName(name);
 
         if (studentToDelete == null) {
             throw new CommandException(MESSAGE_NONEXISTENT_STUDENT);
         }
 
+        // if student number is provided, check that it matches the student number of the student to delete
+        if (!studentToDelete.getStudentNumber().equals(this.studentNumber) && this.studentNumber != null) {
+            throw new CommandException(MESSAGE_NONEXISTENT_STUDENT);
+        }
+
```

Comment:
Perhaps you could just start with two branches where studentNumber is `null` or not. That way u could possibly reduce the number of conditional statements.